bee-api/bee-rest-api/src/endpoints/permission.rs

```diff
@@ -1,30 +1,124 @@
 // Copyright 2020-2021 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use crate::endpoints::rejection::CustomRejection;
+use crate::endpoints::{config::route_to_regex, rejection::CustomRejection, storage::StorageBackend, ApiArgsFullNode};
 
-use warp::{reject, Filter, Rejection};
+use auth_helper::jwt::{Claims, JsonWebToken, TokenData};
+use base64;
 
-use std::net::{IpAddr, SocketAddr};
+use lazy_static::lazy_static;
+use regex::RegexSet;
+use warp::{
+    http::header::{HeaderMap, AUTHORIZATION},
+    path::FullPath,
+    reject, Filter, Rejection,
+};
 
-pub fn has_permission(
-    route: &'static str,
-    public_routes: Box<[String]>,
-    allowed_ips: Box<[IpAddr]>,
+use std::sync::Arc;
+
+/// Bearer for JWT. Please note the whitespace " " is important for correct parsing.
+const BEARER: &str = "Bearer ";
+pub const API_AUDIENCE_CLAIM: &str = "api";
+pub const DASHBOARD_AUDIENCE_CLAIM: &str = "dashboard";
+
+lazy_static! {
+    static ref DASHBOARD_ROUTES: RegexSet = {
+        let routes = vec![
+            "/api/v1/info",
+            "/api/v1/messages/*",
+            "/api/v1/outputs/*",
+            "/api/v1/addresses/*",
+            "/api/v1/milestones/*",
+            "/api/v1/peers*",
+        ];
+        RegexSet::new(routes.iter().map(|r| route_to_regex(r)).collect::<Vec<String>>()).unwrap()
+    };
+}
+
+pub(crate) fn check_permission<B: StorageBackend>(
+    args: Arc<ApiArgsFullNode<B>>,
 ) -> impl Filter<Extract = (), Error = Rejection> + Clone {
-    warp::addr::remote()
-        .and_then(move |addr: Option<SocketAddr>| {
-            let route = route.to_owned();
-            let public_routes = public_routes.clone();
-            let allowed_ips = allowed_ips.clone();
+    warp::any()
+        .and(warp::filters::path::full())
+        .and(warp::header::headers_cloned())
+        .and_then(move |path: FullPath, headers| {
+            let args = args.clone();
             async move {
-                if let Some(v) = addr {
-                    if allowed_ips.contains(&v.ip()) || public_routes.contains(&route) {
+                if args.rest_api_config.public_routes.is_match(path.as_str()) {
+                    return Ok(());
+                }
+
+                let jwt = extract_jwt(&headers)?;
+                // Decode the JWT payload to find out how to validate it.
+                let jwt_payload = {
+                    let jwt_string = jwt.to_string();
+                    let split = jwt_string.split('.').collect::<Vec<&str>>();
+                    if split.len() < 3 {
+                        return Err(reject::custom(CustomRejection::Forbidden));
+                    }
+                    let encoded = split[1];
+                    let decoded = base64::decode(encoded).map_err(|_| reject::custom(CustomRejection::Forbidden))?;
+                    String::from_utf8(decoded).map_err(|_| CustomRejection::Forbidden)?
```

Comment:
I would maybe comment this a little (and add some newlines) just to explain what is going on here/why these operations are necessary.

Reply:
Added some more comments to the decoding process!